File: tools/b2d_h7e_admin_e2e.js
#!/usr/bin/env node
/* Batch H7e — the admin side of the batch: the Factory Information subpage.
 *
 * WHAT THIS HAS TO SHOW, AND WHY THE OBVIOUS ASSERTION IS NOT ENOUGH
 *
 * The batch adds a settings page. "The page renders" is nearly worthless as a
 * check, because the page is new: there is nothing for it to disagree with. So
 * the run is built around three claims that CAN fail:
 *
 *   1. WHICH THEME IS SERVING wp-admin. Every request carries
 *      X-SF-Preflight: 1, so the copy under test is the preflight theme — but
 *      the new page deliberately loads no stylesheet and no script of the
 *      theme's, so on THIS page there is no asset URL to read the answer off.
 *      The contrast page fixes that: the Container Library does carry
 *      sf-mb.css and the two shared scripts, so the run asserts their URL is
 *      the preflight copy's, and it asserts that on the page this batch added
 *      they are ABSENT. Without the first half the second is satisfied by a
 *      site where the assets are broken everywhere.
 *   2. The two fields exist, are text inputs, and carry the DEFAULTS THE BATCH
 *      CLAIMS — 'Linyi, Shandong, China' and 'Available', character for
 *      character. This is the only place either string is visible as a field
 *      rather than as rendered output.
 *   3. The form posts to the Site Settings option group. A page that rendered
 *      perfectly and posted to a group nobody registered would look fine here
 *      and silently not save.
 *
 * A fourth thing is not asserted because it cannot be: that the page is not in
 * the live theme. A submenu added by this batch does not exist on the live
 * copy, so the page simply would not resolve — which makes "it resolved and
 * rendered the right values" the assertion, carried by the values themselves.
 *
 * Nothing is saved. The form is read, never submitted: a settings save is a DB
 * write and this batch does not make one. The sanitizer's empty-falls-back
 * behaviour is asserted in the source pass, against the same code the parent
 * page uses.
 *
 * usage: NODE_PATH=<workspace>/node_modules node tools/b2d_h7e_admin_e2e.js
 *        (password read from /tmp/sf-e2e-pass.txt)
 */
const { chromium } = require('playwright-core');
const fs = require('fs');

const BASE = 'https://dev.zxpet.com';
const PASS = fs.readFileSync('/tmp/sf-e2e-pass.txt', 'utf8').trim();
const OUT = '/Users/meng/WorkBuddy/sinofresh外贸网站建设/docs/batchH7e-shots';
const ORIGIN = 'Linyi, Shandong, China';
const OEM = 'Available';
fs.mkdirSync(OUT, { recursive: true });

(async () => {
  const fails = [];
  const ck = (ok, label) => {
    console.log((ok ? 'PASS' : 'FAIL') + ' ' + label);
    if (!ok) fails.push(label);
  };

  const browser = await chromium.launch({ headless: true });
  const ctx = await browser.newContext({
    viewport: { width: 1440, height: 1080 },
    deviceScaleFactor: 2,
    httpCredentials: { username: 'sfdev', password: 'VkEws18Kl5V1qp3TpZ6s' },
    extraHTTPHeaders: { 'X-SF-Preflight': '1' },
  });
  const page = await ctx.newPage();
  const jsErrors = [];
  page.on('pageerror', e => jsErrors.push('[' + page.url().slice(-44) + '] ' + String(e).slice(0, 150)));

  // ---------------------------------------------------------------- login
  // Retried, and diagnosable. The first version was a single click plus a
  // 30-second waitForURL, and when that timed out the run said only
  // "FATAL page.waitForURL: Timeout 30000ms exceeded" — which is the same
  // message for a rejected password, a redirect somewhere unexpected, and a
  // login page that never got the POST. Three attempts with the landing URL,
  // the page title and wp-login's own error box printed on each miss cost
  // nothing and turn that into an answer.
  let logged = false;
  for (let attempt = 1; attempt <= 3 && !logged; attempt++) {
    await page.goto(BASE + '/wp-login.php', { waitUntil: 'domcontentloaded', timeout: 60000 });
    if (/\/wp-admin\//.test(page.url())) { logged = true; break; }   // cookie already good
    await page.fill('#user_login', 'sf-e2e');
    await page.fill('#user_pass', PASS);
    await page.click('#wp-submit').catch(() => {});
    try {
      await page.waitForURL(u => /\/wp-admin\//.test(u.href), { timeout: 40000 });
      logged = true;
    } catch (e) {
      const diag = await page.evaluate(() => ({
        url: location.href,
        title: document.title,
        err: (document.querySelector('#login_error') || {}).innerText || '',
        body: (document.body.innerText || '').replace(/\s+/g, ' ').slice(0, 220),
      })).catch(() => ({}));
      console.log('  login attempt ' + attempt + ' did not land: ' + JSON.stringify(diag));
      await page.waitForTimeout(2500);
    }
  }
  ck(logged, 'logged in as the throwaway admin');
  if (!logged) { await browser.close(); process.exit(3); }

  // ------------------------------------------- the contrast: a page that
  // DOES carry the shared admin assets, so that "absent" below means absent
  await page.goto(BASE + '/wp-admin/admin.php?page=sf-containers', { waitUntil: 'domcontentloaded' });
  await page.waitForTimeout(1500);
  const contrast = await page.evaluate(() => ({
    title: (document.querySelector('.wrap h1') || {}).textContent || '',
    mb: [...document.querySelectorAll('link[href*="sf-mb.css"]')].map(e => e.href),
    tables: [...document.querySelectorAll('script[src*="sf-mb-tables.js"]')].map(e => e.src),
    settings: [...document.querySelectorAll('script[src*="sf-site-settings.js"]')].map(e => e.src),
    live: [...document.querySelectorAll('link[href*="themes/sinofresh-theme/"], script[src*="themes/sinofresh-theme/"]')].length,
  }));
  ck(contrast.title === 'Container Library', 'the contrast page is the Container Library');
  ck(contrast.mb.length === 1, 'it carries the shared admin stylesheet: ' + (contrast.mb[0] || 'none'));
  ck(/\?ver=/.test(contrast.mb[0] || ''), '...with a version token on it');
  ck((contrast.mb[0] || '').includes('sinofresh-theme-preflight'),
    '...served by the PREFLIGHT copy, which is what makes "which theme" a fact');
  ck(contrast.tables.length === 1 && contrast.settings.length === 1,
    'and both shared scripts are on it too');
  ck(contrast.live === 0, 'no live-theme admin asset URL anywhere on either page');
  await page.screenshot({ path: OUT + '/h7e-03-container-library-contrast.png', fullPage: true });

  // ------------------------------------------------- the new subpage
  await page.goto(BASE + '/wp-admin/admin.php?page=sf-factory-info', { waitUntil: 'domcontentloaded' });
  await page.waitForTimeout(1500);
  const fi = await page.evaluate(() => {
    const q = s => document.querySelector(s);
    const menu = [...document.querySelectorAll('#toplevel_page_sf-site-settings .wp-submenu a')]
      .map(a => a.textContent.trim()).filter(Boolean);
    const cur = q('#toplevel_page_sf-site-settings .wp-submenu a.current')
      || q('#toplevel_page_sf-site-settings .wp-submenu li.current a');
    return {
      url: location.href,
      h1: (q('.wrap h1') || {}).textContent || '',
      bodyHasDenied: /not allowed to access this page/i.test(document.body.innerText),
      origin: (q('#sf_factory_origin') || {}).value,
      originName: (q('#sf_factory_origin') || {}).name,
      originType: (q('#sf_factory_origin') || {}).type,
      oem: (q('#sf_factory_oem') || {}).value,
      oemName: (q('#sf_factory_oem') || {}).name,
      oemType: (q('#sf_factory_oem') || {}).type,
      desc: [...document.querySelectorAll('.wrap .description')].map(e => e.textContent.trim()),
      labels: [...document.querySelectorAll('.wrap th[scope="row"]')].map(e => e.textContent.trim()),
      optionPage: (q('input[name="option_page"]') || {}).value || '',
      nonce: !!q('input[name="_wpnonce"]'),
      submit: (q('#submit') || {}).value || '',
      // the batch's own claim: this page needs none of the shared assets
      mb: document.querySelectorAll('link[href*="sf-mb.css"]').length,
      tables: document.querySelectorAll('script[src*="sf-mb-tables.js"]').length,
      settings: document.querySelectorAll('script[src*="sf-site-settings.js"]').length,
      media: typeof (window.wp && window.wp.media) !== 'undefined',
      menu,
      current: cur ? cur.textContent.trim() : '',
    };
  });

  ck(!fi.bodyHasDenied, 'the subpage resolves (a submenu this batch adds does not exist on the live copy)');
  ck(fi.h1 === 'Factory Information', 'its heading is Factory Information: ' + JSON.stringify(fi.h1));
  ck(fi.origin === ORIGIN, 'Place of Origin defaults to the constant it replaced: ' + JSON.stringify(fi.origin));
  ck(fi.oem === OEM, 'OEM / ODM defaults to the constant it replaced: ' + JSON.stringify(fi.oem));
  ck(fi.originType === 'text' && fi.oemType === 'text', 'both fields are text inputs');
  ck(fi.originName === 'sf_factory_origin' && fi.oemName === 'sf_factory_oem',
    'and they post under the option names the reader reads: ' + fi.originName + ' / ' + fi.oemName);
  ck(fi.labels.join('|') === 'Place of Origin|OEM / ODM',
    'the labels name the two spec-sheet rows: ' + JSON.stringify(fi.labels));
  ck(fi.desc.length === 2, 'each field explains which row it feeds (' + fi.desc.length + ' descriptions)');
  ck(fi.optionPage === 'sf_site_settings',
    'the form posts to the Site Settings option group: ' + JSON.stringify(fi.optionPage));
  ck(fi.nonce, 'and carries a settings nonce');
  ck(/^Save Changes$/i.test(fi.submit) || fi.submit !== '', 'the submit button is present: ' + JSON.stringify(fi.submit));

  // the declared absence, now meaningful because the contrast page had them
  ck(fi.mb === 0, 'the new page loads NO theme stylesheet (sf-mb.css absent)');
  ck(fi.tables === 0, '...and none of the row-table script');
  ck(fi.settings === 0, '...and not the settings row helper either');
  ck(fi.media === false, '...and wp.media is not enqueued for it');

  ck(fi.menu.includes('Container Library') && fi.menu.includes('Global FAQ')
     && fi.menu.includes('Factory Information'),
    'Site Settings now offers three subpages: ' + JSON.stringify(fi.menu));
  ck(fi.current === 'Factory Information', '...and this one is the current menu item: ' + JSON.stringify(fi.current));

  await page.screenshot({ path: OUT + '/h7e-01-factory-info.png', fullPage: true });
  const box = await page.locator('.wrap form').boundingBox();
  if (box) {
    await page.screenshot({
      path: OUT + '/h7e-02-factory-info-fields.png',
      clip: { x: Math.max(0, box.x - 24), y: Math.max(0, box.y - 24),
              width: Math.min(1440, box.width + 48), height: Math.min(1080, box.height + 48) },
    });
    ck(true, 'field-cluster frame written');
  } else {
    ck(false, 'the form could not be located for the close frame');
  }

  ck(jsErrors.length === 0, 'zero page JS errors across every visited screen');
  if (jsErrors.length) jsErrors.slice(0, 6).forEach(e => console.log('    ' + e));

  await browser.close();
  console.log('\n' + (fails.length ? 'FAIL ' + fails.length + ' check(s)' : 'PASS') + '  H7e admin E2E');
  if (fails.length) { console.log('  ' + fails.join('\n  ')); process.exit(1); }
})().catch(e => { console.error('FATAL ' + e.message); process.exit(2); });

File: tools/b2d_h7e_admin_e2e.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Batch H7e — orchestrator for the admin half of the browser pass.

The batch adds a settings page, and a settings page can only be checked where
it lives: in wp-admin, behind a login, with the preflight copy serving the
theme. That needs a user, and a user is a DB row, so the run is built the way
batch H2a built its admin pass — throwaway credentials, deleted afterwards, with
the user list compared before and after rather than assumed.

Sequence:
  1. read the user list                                    -> before
  2. create sf-e2e with a random password (never a fixed one)
  3. run tools/b2d_h7e_admin_e2e.js against the candidate (X-SF-Preflight: 1)
  4. delete sf-e2e
  5. read the user list again                              -> after
  6. assert before == after, the password file is gone, and mu-plugins/ holds
     exactly what it held — the probe that ran before this one installs a
     fixture there, and a fixture left behind is a filter left behind

Nothing here writes to the theme and nothing is saved from wp-admin: the E2E
reads the form and never submits it. The only DB writes are the two user rows
the throwaway account costs, and they are removed in step 4.
"""

import argparse
import json
import os
import secrets
import string
import subprocess
import sys

HOST = 'root@65.49.215.152'
PUB = '/var/www/dev.zxpet.com/public'
MU = PUB + '/wp-content/mu-plugins'
WP = ('cd /var/www/dev.zxpet.com/site-repo && wp --allow-root '
      '--path=' + PUB)
ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
NODE = '/Users/meng/.workbuddy/binaries/node/versions/22.22.2-3/bin/node'
NODE_PATH = '/Users/meng/.workbuddy/binaries/node/workspace/node_modules'
JS = os.path.join(ROOT, 'tools', 'b2d_h7e_admin_e2e.js')
PASSFILE = '/tmp/sf-e2e-pass.txt'
LOGIN = 'sf-e2e'


def wp(args, timeout=180):
    return subprocess.run(['ssh', '-o', 'ConnectTimeout=25', HOST, WP + ' ' + args],
                          capture_output=True, text=True, timeout=timeout)


def user_list():
    r = wp('user list --field=user_login --format=csv')
    if r.returncode != 0:
        raise SystemExit('FATAL wp user list: %s' % (r.stderr or r.stdout)[:300])
    return [l.strip() for l in r.stdout.splitlines() if l.strip() and l.strip() != 'user_login']


def mu_list():
    r = subprocess.run(['ssh', '-o', 'ConnectTimeout=25', HOST,
                        'ls -1 %s 2>/dev/null | sort' % MU], capture_output=True, text=True)
    return [l.strip() for l in r.stdout.splitlines() if l.strip()]


def main(argv=None):
    ap = argparse.ArgumentParser()
    ap.add_argument('--json')
    args = ap.parse_args(argv)

    rows = []

    def ok(label, cond, detail=''):
        rows.append({'label': label, 'ok': bool(cond), 'detail': str(detail)})
        print('  %s  %-58s %s' % ('ok  ' if cond else 'FAIL', label, detail))
        return bool(cond)

    print('== H7e admin E2E ==')

    before = user_list()
    mu_before = mu_list()
    ok('the user list is readable', len(before) > 0, '%d user(s)' % len(before))
    ok('no throwaway admin is left over from an earlier run',
       LOGIN not in before, 'login=%s' % LOGIN)
    if LOGIN in before:
        print('  removing the stale one first')
        wp('user delete %s --yes --reassign=1' % LOGIN)
        before = user_list()

    alphabet = string.ascii_letters + string.digits
    pw = ''.join(secrets.choice(alphabet) for _ in range(24))
    r = wp("user create %s %s@example.com --role=administrator --user_pass='%s' --porcelain"
           % (LOGIN, LOGIN, pw))
    uid = (r.stdout or '').strip().splitlines()[-1].strip() if r.returncode == 0 else ''
    ok('the throwaway admin was created', uid.isdigit(), 'id=%s' % (uid or (r.stderr or '')[:120]))
    if not uid.isdigit():
        return 2

    open(PASSFILE, 'w', encoding='utf-8').write(pw + '\n')
    ok('its password was written to the file the E2E reads',
       os.path.exists(PASSFILE), PASSFILE)

    env = dict(os.environ, NODE_PATH=NODE_PATH)
    p = subprocess.run([NODE, JS], capture_output=True, text=True, env=env, timeout=600)
    tail = (p.stdout or '').strip().splitlines()
    for line in tail[-40:]:
        print('    ' + line)
    if p.stderr.strip():
        print('    stderr: ' + p.stderr.strip()[:400])
    ok('the admin E2E passes against the candidate', p.returncode == 0,
       'exit=%d' % p.returncode)

    r = wp('user delete %s --yes --reassign=1' % uid)
    ok('the throwaway admin was deleted', r.returncode == 0,
       (r.stdout + r.stderr).strip()[:120])

    # The password file lives HERE — on the machine that generated the password
    # and the only machine the browser reads it from — so it has to be removed
    # HERE, and the check has to look at the same path it wrote. The first
    # version of this orchestrator wrote it locally, removed it over ssh, and
    # then read the LOCAL path: the one cleanup step that actually matters, the
    # only place the throwaway password exists, was reported as failed while the
    # file sat on disk, and the remote delete was a no-op against a path that
    # never existed there. Both halves are now separate claims.
    try:
        os.remove(PASSFILE)
    except FileNotFoundError:
        pass
    ok('the password file is gone from this machine', not os.path.exists(PASSFILE), PASSFILE)
    r = subprocess.run(['ssh', '-o', 'ConnectTimeout=20', HOST,
                        'test -e %s && echo PRESENT || echo GONE' % PASSFILE],
                       capture_output=True, text=True)
    ok('...and no copy was left on the dev box either', 'GONE' in r.stdout, '')

    after = user_list()
    ok('the user list is exactly what it was (zero DB trace)',
       before == after, 'before=%d after=%d' % (len(before), len(after)))

    mu_after = mu_list()
    ok('mu-plugins/ holds exactly what it held (no fixture left behind)',
       mu_before == mu_after, '%s' % mu_after)
    ok('...and the H7e probe fixture is not among them',
       not any('h7e' in f for f in mu_after), '')

    passed = all(r['ok'] for r in rows)
    print('\n%s  H7e admin E2E' % ('PASS' if passed else 'FAIL'))
    if args.json:
        json.dump({'ok': passed, 'rows': rows}, open(args.json, 'w'), indent=1)
        print('  json -> %s' % args.json)
    return 0 if passed else 1


if __name__ == '__main__':
    sys.exit(main())
